File: apps/backend/app/models/theme.py
from typing import Optional

from beanie import Document
from pydantic import BaseModel, ConfigDict, Field, model_validator
from bson.objectid import ObjectId
from .user import PyObjectId


_CAMEL_TO_SNAKE = {
    "primaryColor": "primary_color",
    "secondaryColor": "secondary_color",
    "textColor": "text_color",
    "headerText": "header_text",
    "inputPlaceholder": "input_placeholder",
    "borderRadius": "border_radius",
    "showHeader": "show_header",
}


class ThemeUpdate(BaseModel):
    """Payload for creating/updating a theme. All fields optional."""

    model_config = ConfigDict(extra="ignore")

    theme: Optional[str] = None
    position: Optional[str] = None
    primary_color: Optional[str] = None
    secondary_color: Optional[str] = None
    text_color: Optional[str] = None
    header_text: Optional[str] = None
    input_placeholder: Optional[str] = None
    width: Optional[str] = None
    height: Optional[str] = None
    border_radius: Optional[str] = None
    launcher: Optional[bool] = None
    show_header: Optional[bool] = None

    @model_validator(mode="before")
    @classmethod
    def accept_camel_case_keys(cls, data):
        if not isinstance(data, dict):
            return data
        return {_CAMEL_TO_SNAKE.get(key, key): value for key, value in data.items()}


class Theme(Document):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    theme: str = Field(default="light")
    position: str = Field(default="bottom-right")
    primary_color: str = Field(default="#3B82F6")
    secondary_color: str = Field(default="#F3F4F6")
    text_color: str = Field(default="#000000")
    header_text: str = Field(default="Chat with me")
    input_placeholder: str = Field(default="Type your message here...")
    width: str = Field(default="300px")
    height: str = Field(default="500px")
    border_radius: str = Field(default="8px")
    launcher: bool = Field(default=True)
    show_header: bool = Field(default=True)

    class Config:
        json_encoders = {ObjectId: str}

    class Settings:
        name = "theme_config"

File: apps/frontend/src/lib/api/endpoints/theme.ts
import { apiClient } from "./client";
import type { ThemeConfig } from "@/types/config";

type ThemePayload = {
    theme?: string;
    position?: string;
    primary_color?: string;
    secondary_color?: string;
    text_color?: string;
    header_text?: string;
    input_placeholder?: string;
    width?: string;
    height?: string;
    border_radius?: string;
    launcher?: boolean;
    show_header?: boolean;
}

function toThemeConfig(payload: ThemePayload): ThemeConfig {
    return {
        theme: payload.theme ?? "light",
        position: payload.position ?? "bottom-right",
        primaryColor: payload.primary_color ?? "#3B82F6",
        secondaryColor: payload.secondary_color ?? "#F3F4F6",
        textColor: payload.text_color ?? "#000000",
        headerText: payload.header_text ?? "Chat with me",
        inputPlaceholder: payload.input_placeholder ?? "Type your message here...",
        width: payload.width ?? "350px",
        height: payload.height ?? "500px",
        borderRadius: payload.border_radius ?? "8px",
        launcher: payload.launcher ?? true,
        showHeader: payload.show_header ?? true,
    }
}

function toThemePayload(config: ThemeConfig): ThemePayload {
    return {
        theme: config.theme,
        position: config.position,
        primary_color: config.primaryColor,
        secondary_color: config.secondaryColor,
        text_color: config.textColor,
        header_text: config.headerText,
        input_placeholder: config.inputPlaceholder,
        width: config.width,
        height: config.height,
        border_radius: config.borderRadius,
        launcher: config.launcher,
        show_header: config.showHeader,
    }
}


// TODO: workspaceid cannot be undefined
export const themeAPI = {
    getTheme: (workspaceId: string | undefined): Promise<ThemeConfig> =>
        apiClient.get<ThemePayload>(`/theme?workspace_id=${workspaceId}`).then((res) => toThemeConfig(res.data)),
    createTheme: (workspaceId: string | undefined, theme: ThemeConfig): Promise<ThemeConfig> =>
        apiClient.post<ThemePayload>(`/theme?workspace_id=${workspaceId}`, toThemePayload(theme)).then((res) => toThemeConfig(res.data)),
    updateTheme: (workspaceId: string | undefined, theme: ThemeConfig): Promise<ThemeConfig> =>
        apiClient.put<ThemePayload>(`/theme?workspace_id=${workspaceId}`, toThemePayload(theme)).then((res) => toThemeConfig(res.data)),
}
